horovod/run/run_func.py

```diff
@@ -0,0 +1,199 @@
+                          the initialization checks only once every 60
+                          minutes -- if the checks successfully pass.
+                          Otherwise, all the checks will run every time
+                          horovodrun is called."
+    :param start_timeout: Horovodrun has to perform all the checks and
+                          start the processes before the specified
+                          timeout. The default value is 30 seconds.
+                          Alternatively, The environment variable
+                          HOROVOD_START_TIMEOUT can also be used to
+                          specify the initialization timeout.
+    :param env: Environment dictionary to use in Horovod run.
+                If None, use `os.environ`.
+    :param verbose: If this flag is set, extra messages will be printed.
+
+    :return: A list contains returned result from each MPI process. The ith value of the list
+             corresponds to the returned result of rank i process.
+    """
+
+    wdir = tempfile.mkdtemp()
+    if verbose:
+        print("horovod.run_func working dir is " + wdir)
+
+    if ssh_port:
+        ssh_port_opt = "-p " + str(ssh_port)
+        scp_port_opt = "-P " + str(ssh_port)
+    else:
+        ssh_port_opt = ""
+        scp_port_opt = ""
+
+    def wrapped_proc_fn(rank=0):
+        return_value = fn()
+
+        result_file = "{prefix}.{rank}".format(prefix=_PICKLED_RESULT_PREFIX, rank=rank)
+        with open(result_file, 'wb') as fp:
+            try:
+                cloudpickle.dump(return_value, fp)
+            except Exception as e:
+                raise RuntimeError("Caught an excpetion while pickling "
+                                   "return value: {}".format(type(e)), e)
+
+    pickled_proc_fn_str = cloudpickle.dumps(wrapped_proc_fn)
+    local_pickled_proc_fn_path = os.path.join(wdir, _LOCAL_PICKLED_PROC_FN_FILENAME)
+    with open(local_pickled_proc_fn_path, 'wb') as f:
+        f.write(pickled_proc_fn_str)
+
+    local_launcher_path = os.path.join(wdir, _LOCAL_PROC_LAUNCHER_FILENAME)
+    with open(local_launcher_path, 'w') as f:
+        f.write(textwrap.dedent("""
+        set -e
+        cd {wdir}
+        rank=${{OMPI_COMM_WORLD_RANK:-0}}
+        {exec_path} -c "import cloudpickle; cloudpickle.load(open('{fn_file}', 'rb'))(rank=$rank)"
+        """.format(wdir=wdir,
+                   exec_path=sys.executable,
+                   fn_file=_PICKLED_PROC_FN_FILENAME)))
+
+    if host is None:
+        host = "localhost:{np}".format(np=num_proc)
+
+    all_host_names = set(x.split(':')[0] for x in host.split(','))
+
+    # copy pickled function and launcher script to all hosts via scp
+    # TODO: parallelize the scp copy
+    for hostname in all_host_names:
+        scp_cmd = textwrap.dedent(
+            """
+            ssh {ssh_port_opt} -o StrictHostKeyChecking=no {hostname} mkdir -p {wdir} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
```

Comment:
Current impl will change file name when copy to dest location (remove "local_" prefix). This is to avoid error when scp copy to local machine.
But the scp copy multiple file syntax `scp "/path/fileA /path/fileB" remote:/path/` could not change dest file name.
Another way is to use `scp -r` to copy the whole dir to dest but it generate more dirs.